### PR TITLE
Null safety master fixes pr

### PR DESCRIPTION
**Pull request by fjnoyp**
Fix unit tests issues in Dart and remove redundant exception. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_datastore_plugin_interface/lib/src/types/temporal/temporal_date.dart [hunk 1]
```diff
@@ -74,7 +74,7 @@ class TemporalDate {
 
     // Validate
     String? regexString = regExp.stringMatch(iso8601String);
-    if (regexString != iso8601String) {
+    if (regexString == null || regexString != iso8601String) {
```
[the remark this answers is not in the document]
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/lib/src/types/temporal/temporal_date.dart:77]: @HuiSF we will be merging in null safety changes to the datastore plugin **interface** now but no changes to the datastore plugin. 

This is because the amplify flutter plugin has dependencies on all plugin interfaces but not the actual plugins themselves.

File: packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart [hunk 2]
```diff
@@ -17,18 +17,14 @@ import 'model_schema.dart';
...
-  final String id;
-
   ModelType getInstanceType();
 
-  const Model({required this.id});
-
-  String getId() {
-    return id;
-  }
+  String getId();
```
Comment (Jordan-Nelson) [inline packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart:22]: Should this throw an UnimplementedError if it no longer has a default implementation?

File: packages/amplify_datastore_plugin_interface/test/testData/Blog.dart [hunk 3]
```diff
@@ -13,19 +13,28 @@
...
 class Blog extends Model {
   static const classType = const BlogType();
   final String id;
-  final String name;
-  final List<Post>? posts;
+  final String? _name;
+  final List<Post>? _posts;
+
+  String get name {
+    return _name!;
```
Comment (Jordan-Nelson) [inline packages/amplify_datastore_plugin_interface/test/testData/Blog.dart:32]: We are already discussing this one offline, but I don't see how this could be a safe cast.

File: packages/amplify_datastore_plugin_interface/test/testData/Blog.dart [hunk 4]
```diff
@@ -65,29 +76,32 @@ class Blog extends Model {
     var buffer = new StringBuffer();
 
     buffer.write("Blog {");
-    buffer.write("id=" + id + ", ");
-    buffer.write("name=" + name);
+    buffer.write("id=" + "$id" + ", ");
```
Comment (Jordan-Nelson) [inline packages/amplify_datastore_plugin_interface/test/testData/Blog.dart:79]: *nit:* If you are going to use string interpolation
```dart
buffer.write("id=$id, ");
```

File: packages/amplify_datastore_plugin_interface/test/testData/ModelProvider.dart [hunk 5]
```diff
@@ -27,14 +28,9 @@ export 'PostAuthComplex.dart';
...
 class ModelProvider implements ModelProviderInterface {
   @override
-  String version = "6319a5b46c3ea5e5817b63396bc464bb";
+  String version = "33b1717ef9afb390e3f7431e8f3859fc";
   @override
-  List<ModelSchema> modelSchemas = [
-    Blog.schema,
-    Comment.schema,
-    Post.schema,
-    PostAuthComplex.schema
```
Comment (ragingsquirrel3) [inline packages/amplify_datastore_plugin_interface/test/testData/ModelProvider.dart:36]: question: Why is `PostAuthComplex` omitted now?

File: packages/amplify_datastore_plugin_interface/test/testData/Post.dart [hunk 6]
```diff
@@ -13,22 +13,43 @@
...
 @immutable
 class Post extends Model {
   static const classType = const PostType();
   final String id;
-  final String title;
-  final int? rating;
-  final DateTime? created;
-  final Blog? blog;
-  final List<Comment>? comments;
+  final String? _title;
```
Comment (ragingsquirrel3) [inline packages/amplify_datastore_plugin_interface/test/testData/Post.dart:28]: question: Should `title` be required here? Is it required in `_internal`? Same question for `rating`? I'm just a little confused by different syntax for accessors below but I'm likely missing something.

File: packages/amplify_datastore_plugin_interface/test/testData/PostAuthComplex.dart [hunk 7]
```diff
@@ -21,8 +21,16 @@ import 'package:flutter/foundation.dart';
 class PostAuthComplex extends Model {
   static const classType = const PostAuthComplexType();
   final String id;
-  final String title;
-  final String? owner;
+  final String? _title;
+  final String? _owner;
+
+  String get title {
```
Comment (ragingsquirrel3) [inline packages/amplify_datastore_plugin_interface/test/testData/PostAuthComplex.dart:27]: Again, I guess I'm a little confused as `_title` is optional but not sure why this does not have `String?` syntax like owner below.

File: packages/amplify_datastore_plugin_interface/test/testData/Blog.dart [hunk 8]
```diff
@@ -65,29 +76,32 @@ class Blog extends Model {
...
+    buffer.write("name=" + "$_name");
```
Comment (HuiSF) [inline packages/amplify_datastore_plugin_interface/test/testData/Blog.dart:80]: nit: `buffer.write("name=$_name");`

File: packages/amplify_datastore_plugin_interface/test/testData/Post.dart [hunk 9]
```diff
@@ -1,5 +1,5 @@
 /*
-* Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+* Copyright 201 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (ragingsquirrel3) [inline packages/amplify_datastore_plugin_interface/test/testData/Post.dart:2]: nit: Is this change intentional?

File: packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart (hunk 2 above)
Comment (ragingsquirrel3) [inline packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart:22]: question: Why continue to include this method written this way? Extended classes will continue to use @override, correct? Not requesting a change, just trying to understand.

**Review by ragingsquirrel3 — COMMENTED** [review]
Overall, looks good, just nits/questions to help me understand. Likely I can approve if I understand correctly, just trying to make sure.

File: packages/amplify_datastore_plugin_interface/test/testData/Comment.dart [hunk 10]
```diff
@@ -79,13 +90,13 @@ class Comment extends Model {
 
   Comment.fromJson(Map<String, dynamic> json)
       : id = json['id'],
-        post = json['post'] != null
+        _post = json['post'] != null
             ? Post.fromJson(new Map<String, dynamic>.from(json['post']))
             : null,
-        content = json['content'];
+        _content = json['content'];
```
Comment (HuiSF) [inline packages/amplify_datastore_plugin_interface/test/testData/Comment.dart:96]: Should we validate `json['content']` here to ensure it's non-null? Asking because that it looks like class `Comment` ensures `_content` cannot be assigned `null`, except this constructor. Thinking it may be better to keep the behavior consistent. (same question to other `fromJson` methods in other classes)

File: packages/amplify_datastore_plugin_interface/test/testData/Post.dart (hunk 9 above)
Comment (HuiSF) [inline packages/amplify_datastore_plugin_interface/test/testData/Post.dart:2]: nit: `2021`

File: packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart (hunk 2 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart:22]: Not a major concern as all models will be generated by codegen team which is guaranteed to have that implementation.

File: packages/amplify_datastore_plugin_interface/test/testData/Blog.dart (hunk 3 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/test/testData/Blog.dart:32]: Yup it's not.  The Blog graphqL schema has a required name field.  Customers wouldn't expect a required field to be nullable.  However, we make the underlying private field nullable to handle the fact that the model might not have that field if it was a nested model with only the id field set.

File: packages/amplify_datastore_plugin_interface/test/testData/Blog.dart (hunk 4 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/test/testData/Blog.dart:79]: good call. thanks!

File: packages/amplify_datastore_plugin_interface/test/testData/ModelProvider.dart (hunk 5 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/test/testData/ModelProvider.dart:36]: good call.  it shouldn't ... added it back in

File: packages/amplify_datastore_plugin_interface/test/testData/Post.dart (hunk 6 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/test/testData/Post.dart:28]: the codegen model fields are generated based on a graphql schema that unfortunately isn't shown in the file diff here.  The underlying fields (private _) are all nullable because we might create a model with just an id.  However, the field can be marked as required in the graphql schema.  Thus the public accessor makes a force cast "!"

File: packages/amplify_datastore_plugin_interface/test/testData/PostAuthComplex.dart (hunk 7 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/test/testData/PostAuthComplex.dart:27]: see answer above.

File: packages/amplify_datastore_plugin_interface/test/testData/Post.dart (hunk 9 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/test/testData/Post.dart:2]: Nope thanks for catching!

File: packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart (hunk 2 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart:22]: We need to include this method since Model is an abstract class.  And abstract class needs to have method so we can call it without need to cast to the underlying Model type like Blog, Post, etc.

File: packages/amplify_datastore_plugin_interface/test/testData/Blog.dart (hunk 8 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/test/testData/Blog.dart:80]: Thanks!

File: packages/amplify_datastore_plugin_interface/test/testData/Comment.dart (hunk 10 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/test/testData/Comment.dart:96]: _content can be null as it is a nullable type.   Typically the field should not be null, but in fromJson we don't know if all fields are going to be set as it might be a nested model in which only the id field will be set.

File: packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart (hunk 2 above)
Comment (Jordan-Nelson) [reply packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart:22]: Yeah, I consider this to be more of a nit/non-blocking comment. It is hard to imagine a scenario where this would actually be an issue, but if there were some bug in the code generation failing fast by throwing an error would be preferred to returning null.

File: packages/amplify_datastore_plugin_interface/test/testData/Blog.dart (hunk 3 above)
Comment (Jordan-Nelson) [reply packages/amplify_datastore_plugin_interface/test/testData/Blog.dart:32]: I think it is fine for now since datastore won't be in the initial preview and since this looks like it is just for tests. I assume there will be more discussion about datastore offline.

File: packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart (hunk 2 above)
Comment (ragingsquirrel3) [reply packages/amplify_datastore_plugin_interface/lib/src/types/models/model.dart:22]: thanks

File: packages/amplify_datastore_plugin_interface/test/testData/Post.dart (hunk 6 above)
Comment (ragingsquirrel3) [reply packages/amplify_datastore_plugin_interface/test/testData/Post.dart:28]: thanks

**Review by HuiSF — APPROVED** [review]
Thanks for resolving/answering comments

**Review by ragingsquirrel3 — APPROVED** [review]
LGTM, thanks for info on questions I had.

**Review by Jordan-Nelson — APPROVED** [review]
Approving, but noting that there needs to be more discussion about the nullable/non-nullable fields in the generate models for datastore.
